### DISCUSSION TITLE
Curl not working from inside a container?

### DISCUSSION THREAD

**Discussion by Emporea** [nginx-proxy/nginx-proxy#1896] [Q&A] [answered]
Lets have a simple webserver printing `OK` when accessed.

Following commands are from host: (this is working)
---
ping ok.example.tld
```
PING ok.example.tld (xxx.xxx.xxx.xxx) 56(84) bytes of data.
64 bytes from hostname (xxx.xxx.xxx.xxx): icmp_seq=1 ttl=64 time=0.053 ms
```
curl -v -L ok.example.tld
```
*   Trying xxx.xxx.xxx.xxx...
* TCP_NODELAY set
* Connected to ok.example.tld (xxx.xxx.xxx.xxx) port 80 (#0)
> GET / HTTP/1.1
> Host: ok.example.tld
> User-Agent: curl/7.68.0
> Accept: */*
OK
```

And now the same commands from inside a docker container. (ping works, curl is stuck forever)
---
ping ok.example.tld
```
PING ok.example.tld (xxx.xxx.xxx.xxx) 56(84) bytes of data.
64 bytes from hostname (xxx.xxx.xxx.xxx): icmp_seq=1 ttl=64 time=0.053 ms
```
curl -v -L ok.example.tld
```
*   Trying xxx.xxx.xxx.xxx:80...
* TCP_NODELAY set
 ```

One can see the very low latency of ping, this is because the domain / public ip gets resolved to the local address cause.

Do you understand what I am trying to do ?
( I need this because my nextcloud container needs to be able to communicate to collabora office via office.example.tld but this fails )

Is this a problem of nginx-proxy, my Server configuration, my A/AAAA Records, docker or a totally different thing?

**Comment by tkw1536** [comment]
I've had this problem before when the hostname of a container was equal to an fqdn. 

For instance, when I have a container called "tld.example.com", then any container on the same network as this container will resolve that name as the internal IP address as that domain. Pinging that IP will work, but curling it will attempt to reach that container on port 80 / 443 (which may or may not work, depending on the exact container). 

This is not really an nginx-proxy specific problem, it's a generic one  The workaround I use is to not name any container as an fqdn, instead I usually append some kind of surffix. For instance instead of naming it "tld.example.com" I name the container "tld.example.com_server" which solves the problem. 

Another (curl-specific) workaround might be to tell curl to use a specific DNS server, or hard-code the DNS results:

```bash
# use 10.0.0.1:443 for tld.example.com
curl --resolve tld.example.com:443:10.0.0.1 https://tld.example.com
# or resolve using public google DNS @ 8.8.8.8
curl --dns-servers 8.8.8.8 https://tld.example.com
```

**Comment by Emporea** [comment, marked as answer]
I solved it... a little bit embarrasing tho.

I didnt allow 80/443 from any other adressed than from outside.. changes this rule and it works.

I forgot it because docker does change iptables and ignores, lets say ufw, anyway.
Appearently this change does not include connections from.within containers on a network!=host

` 0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp   nginx-proxy`
does not allow access from within a custom bridge network appearently..
